File: VoiceRecorder.Audio/FftPitchDetector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace VoiceRecorder.Audio
{
    // FFT based pitch detector. seems to work best with block sizes of 4096
    public class FftPitchDetector : IPitchDetector
    {
        private float sampleRate;

        public FftPitchDetector(float sampleRate)
        {
            this.sampleRate = sampleRate;
        }

        // http://en.wikipedia.org/wiki/Window_function
        private float HammingWindow(int n, int N) 
        {
            return 0.54f - 0.46f * (float)Math.Cos((2 * Math.PI * n) / (N - 1));
        }

        private float[] fftBuffer;
        public float DetectPitch(float[] buffer, int frames)
        {
            Func<int, int, float> window = HammingWindow;
            if (fftBuffer == null)
            {
                fftBuffer = new float[frames * 2];
            }
            for (int n = 0; n < frames; n++)
            {
                fftBuffer[n * 2] = buffer[n]; // *window(n, frames);
                fftBuffer[n * 2 + 1] = 0; // need to clear out as fft modifies buffer
            }

            // assuming frames is a power of 2
            SmbPitchShift.smbFft(fftBuffer, frames, -1);

            float binSize = sampleRate / frames;
            int minBin = (int)(85 / binSize);
            int maxBin = (int)(300 / binSize);
            float maxIntensity = 0f;
            int maxBinIndex = 0;
            for (int bin = minBin; bin <= maxBin; bin++)
            {
                float intensity = fftBuffer[bin * 2] * fftBuffer[bin * 2] + fftBuffer[bin * 2 + 1] * fftBuffer[bin * 2 + 1];
                if (intensity > maxIntensity)
                {
                    maxIntensity = intensity;
                    maxBinIndex = bin;
                }
            }
            return binSize * maxBinIndex;
        }
    }
}

File: ~VB Port/VoiceRecorder.Audio/FftPitchDetector.vb
Imports System.Text

Namespace VoiceRecorder.Audio
    ' FFT based pitch detector. seems to work best with block sizes of 4096
    Public Class FftPitchDetector
        Implements IPitchDetector
        Private sampleRate As Single

        Public Sub New(ByVal sampleRate As Single)
            Me.sampleRate = sampleRate
        End Sub

        ' http://en.wikipedia.org/wiki/Window_function
        Private Function HammingWindow(ByVal n As Integer, ByVal _N As Integer) As Single
            Return 0.54F - 0.46F * CSng(Math.Cos((2 * Math.PI * n) / (_N - 1)))
        End Function

        Private fftBuffer() As Single
        Private prevBuffer() As Single

        Public Function DetectPitch(ByVal buffer() As Single,
                                    ByVal inFrames As Integer) As Single Implements IPitchDetector.DetectPitch
            Dim window As Func(Of Integer, Integer, Single) = AddressOf HammingWindow
            If prevBuffer Is Nothing Then
                prevBuffer = New Single(inFrames - 1) {}
            End If

            ' double frames since we are combining present and previous buffers
            Dim frames = inFrames * 2
            If fftBuffer Is Nothing Then
                fftBuffer = New Single(frames * 2 - 1) {} ' times 2 because it is complex input
            End If

            For n = 0 To frames - 1
                If n < inFrames Then
                    fftBuffer(n * 2) = prevBuffer(n) * window(n, frames)
                    fftBuffer(n * 2 + 1) = 0 ' need to clear out as fft modifies buffer
                Else
                    fftBuffer(n * 2) = buffer(n - inFrames) * window(n, frames)
                    fftBuffer(n * 2 + 1) = 0 ' need to clear out as fft modifies buffer
                End If
            Next n

            ' assuming frames is a power of 2
            SmbPitchShift.smbFft(fftBuffer, frames, -1)

            Dim binSize = sampleRate / frames
            Dim minBin = CInt(Fix(85 / binSize))
            Dim maxBin = CInt(Fix(300 / binSize))
            Dim maxIntensity = 0.0F
            Dim maxBinIndex = 0
            For bin = minBin To maxBin
                Dim real = fftBuffer(bin * 2)
                Dim imaginary = fftBuffer(bin * 2 + 1)
                Dim intensity = real * real + imaginary * imaginary
                If intensity > maxIntensity Then
                    maxIntensity = intensity
                    maxBinIndex = bin
                End If
            Next bin
            Return binSize * maxBinIndex
        End Function
    End Class
End Namespace
